File: validation/learning-site/public-3f72ebc30128c2a3171951050a566d3ed8db7c16/verify-public.mjs
// Read-only bounded verification of the published documentation, not its corpus.
import assert from 'node:assert/strict';
import { createHash } from 'node:crypto';
import { readFile, open } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
const commit = '3f72ebc30128c2a3171951050a566d3ed8db7c16';
const base = 'https://chris-page-gov.github.io/okf-dwp/';
const expectedUrl = new URL('./expected-site-manifest.json', import.meta.url);
const output = process.argv[2];
assert.ok(output, 'Supply a new output JSON path; existing observations are refused');
const handle = await open(output, 'wx');
const hash = raw => createHash('sha256').update(raw).digest('hex');
const started = new Date().toISOString();
const observations = [];
let transferred = 0;
let failure = null;
let expected;
let manifestHash;
const safePath = value => typeof value === 'string' && value.length < 512 && !value.startsWith('/') &&
  !value.includes('\\') && !value.split('/').some(part => !part || part === '..' || part === '.') && !/[?#\x00-\x1f]/.test(value);
async function fetchChecked(relative, expectedBytes, expectedHash) {
  assert.ok(safePath(relative));
  assert.ok(Number.isInteger(expectedBytes) && expectedBytes >= 0 && expectedBytes <= 1024 * 1024);
  const requested = new URL(relative, base).href;
  const began = performance.now();
  const row = { path: relative, url: requested, observed_at: new Date().toISOString() };
  observations.push(row);
  try {
    const response = await fetch(requested, { redirect: 'manual', signal: AbortSignal.timeout(20000), headers: { Accept: '*/*' } });
    row.http_status = response.status; row.response_url = response.url;
    row.content_type = response.headers.get('content-type');
    row.etag = response.headers.get('etag'); row.last_modified = response.headers.get('last-modified');
    assert.equal(response.status, 200, relative + ' HTTP status');
    assert.equal(response.url, requested, 'Canonical URL');
    const reader = response.body.getReader(), parts = [];
    let size = 0;
    while (true) {
      const { done, value } = await reader.read(); if (done) break;
      size += value.byteLength; transferred += value.byteLength;
      if (size > 1024 * 1024 || transferred > 16 * 1024 * 1024) {
        await reader.cancel(); throw new Error('Observed response byte bound exceeded');
      }
      parts.push(Buffer.from(value));
    }
    const raw = Buffer.concat(parts);
    row.bytes = raw.length; row.sha256 = hash(raw);
    assert.equal(raw.length, expectedBytes, relative + ' byte length');
    assert.equal(hash(raw), expectedHash, relative + ' SHA-256');
    row.status = 'matched';
    return raw;
  } catch (error) {
    row.status = 'failed'; row.error = String(error.message || error);
    throw error;
  } finally { row.elapsed_ms = Math.round(performance.now() - began); }
}
try {
  const expectedRaw = await readFile(expectedUrl);
  assert.ok(expectedRaw.length <= 256 * 1024);
  expected = JSON.parse(expectedRaw); manifestHash = hash(expectedRaw);
  assert.equal(expected.schema, 'okf-dwp-learning-site.v1');
  assert.equal(expected.source_commit, commit);
  assert.equal(expected.page_count, 105); assert.equal(expected.source_pages.length, 105);
  assert.equal(expected.files.length, 108);
  assert.equal(new Set(expected.files.map(row => row.path)).size, expected.files.length);
  assert.ok(expected.source_pages.every(row => safePath(row.path) && row.path.endsWith('.md') && !row.path.includes('.email.md')));
  assert.ok(expected.files.every(row => safePath(row.path) && /^[a-f0-9]{64}$/.test(row.sha256) &&
    (row.path.endsWith('.html') || ['assets/learning.css', '.nojekyll'].includes(row.path))));
  assert.ok(expected.files.reduce((sum, row) => sum + row.bytes, 0) <= 16 * 1024 * 1024);
  await fetchChecked('site-manifest.json', expectedRaw.length, manifestHash);
  // Four files at a time. A failed batch fully settles; no later batch starts.
  for (let offset = 0; offset < expected.files.length; offset += 4) {
    const rows = expected.files.slice(offset, offset + 4);
    const batch = await Promise.allSettled(rows.map(row => fetchChecked(row.path, row.bytes, row.sha256)));
    const failed = batch.find(row => row.status === 'rejected');
    if (failed) throw failed.reason;
  }
} catch (error) { failure = String(error.message || error); }
const receipt = { schema: 'okf-dwp-learning-site-public-verification.v1', status: failure ? 'failed' : 'passed',
  started_at: started, observed_at: new Date().toISOString(), source_commit: commit, base_url: base,
  method: 'Actual read-only HTTPS; exact manifest and every listed output compared by byte length and SHA-256; no browser interception or corpus downloads.',
  expected_manifest_sha256: manifestHash, source_pages: expected?.page_count,
  listed_outputs: expected?.files.length, matched_requests: observations.filter(row => row.status === 'matched').length,
  transferred_bytes: transferred, limits: { concurrent_requests: 4, timeout_ms_per_request: 20000,
    max_bytes_per_response: 1048576, max_total_bytes: 16777216, retries: 0 },
  verifier_sha256: hash(await readFile(fileURLToPath(import.meta.url))),
  requests: observations.sort((a,b) => a.path.localeCompare(b.path)), error: failure,
  limitations: ['Point-in-time publication identity check; no availability or performance guarantee.',
    'The 105 source pages produce 106 HTML outputs including the home-page alias, plus a stylesheet and .nojekyll.',
    'No raw correspondence, corpus bodies, external evidence links, browser layout or accessibility acceptance is tested.'] };
await handle.writeFile(JSON.stringify(receipt, null, 2) + '\n'); await handle.close();
console.log(JSON.stringify({ status: receipt.status, source_commit: commit, matched_requests: receipt.matched_requests,
  transferred_bytes: transferred, error: failure }));
if (failure) process.exitCode = 1;

File: validation/learning-site/public-91b9907836c8a340d974dd958969f4f8cbb3a0c6/verify_public.py
#!/usr/bin/env python3
"""Bounded publication identity and internal-link audit; no browser or model calls."""
import argparse
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime, timezone
import hashlib
from html.parser import HTMLParser
import json
import os
from pathlib import Path, PurePosixPath
import posixpath
import stat
import subprocess
import time
from urllib.parse import unquote, urljoin, urlsplit
from urllib.request import HTTPRedirectHandler, Request, build_opener

BASE = "https://chris-page-gov.github.io/okf-dwp/"
COMMIT = "91b9907836c8a340d974dd958969f4f8cbb3a0c6"
MAX_FILES = 200
MAX_MANIFEST = 128 * 1024
MAX_MEMBER = 2 * 1024 * 1024
MAX_TOTAL = 32 * 1024 * 1024
ROOT_DOCS = {"README.md", "CHANGELOG.md", "NOTICE.md", "AI_USAGE.md", "LICENSE_DECISIONS.md", "REPOSITORY_STATUS.md"}


def require(condition, message):
    if not condition:
        raise ValueError(message)


def sha(raw):
    return hashlib.sha256(raw).hexdigest()


def directory(path):
    path = Path(path).absolute()
    for item in [*reversed(path.parents), path]:
        require(stat.S_ISDIR(item.lstat().st_mode), "Directory or parent is not a regular directory")
    return path


def bounded_read(path, limit):
    path = Path(path).absolute()
    directory(path.parent)
    before = path.lstat()
    require(stat.S_ISREG(before.st_mode) and before.st_size <= limit, "Invalid or oversized local file")
    fd = os.open(path, os.O_RDONLY | os.O_NOFOLLOW)
    with os.fdopen(fd, "rb") as stream:
        opened = os.fstat(stream.fileno())
        require((opened.st_dev, opened.st_ino) == (before.st_dev, before.st_ino), "Local file changed before reading")
        raw = stream.read(limit + 1)
    require(len(raw) <= limit and len(raw) == before.st_size == opened.st_size, "Local file size changed")
    return raw


def safe_path(value):
    require(isinstance(value, str) and value and not value.startswith("/") and
            not any(c in value for c in "\\\0?#%") and
            all(part not in {"", ".", ".."} for part in value.split("/")), "Unsafe publication path")
    return value


def validate_manifest(value):
    require(set(value) == {"schema", "source_commit", "source_pages", "page_count", "files", "scope"}, "Unexpected manifest fields")
    require(value["schema"] == "okf-dwp-learning-site.v1" and value["source_commit"] == COMMIT, "Wrong source identity")
    require(isinstance(value["files"], list) and 1 <= len(value["files"]) <= MAX_FILES, "Output census outside bound")
    require(isinstance(value["source_pages"], list) and 1 <= len(value["source_pages"]) <= MAX_FILES, "Source census outside bound")
    require(value["page_count"] == len(value["source_pages"]), "Source page count differs")
    groups = []
    for key in ("source_pages", "files"):
        rows = {}
        for row in value[key]:
            require(isinstance(row, dict) and set(row) == {"path", "bytes", "sha256"}, "Invalid manifest entry")
            name = safe_path(row["path"])
            require(name not in rows, "Duplicate manifest path")
            require(type(row["bytes"]) is int and 0 <= row["bytes"] <= MAX_MEMBER, "Member outside byte bound")
            require(isinstance(row["sha256"], str) and len(row["sha256"]) == 64 and all(c in "0123456789abcdef" for c in row["sha256"]), "Invalid digest")
            if key == "source_pages":
                parts = PurePosixPath(name).parts
                require(name.endswith(".md") and not any(p.startswith(".") for p in parts) and
                        (name in ROOT_DOCS or parts[0] in {"docs", "evaluation"}), "Source is outside public documentation allowlist")
            rows[name] = row
        require(sum(row["bytes"] for row in rows.values()) <= MAX_TOTAL, "Manifest exceeds total byte bound")
        groups.append(rows)
    sources, files = groups
    expected = {str(PurePosixPath(name).with_suffix(".html")) for name in sources} | {"index.html", "assets/learning.css", ".nojekyll"}
    require(set(files) == expected, "Unexpected generated output paths")
    return sources, files


def verify_sources(repo, sources):
    """Only public allowlisted blobs at the exact commit; never the worktree."""
    directory(repo)
    checked = []
    for name, row in sorted(sources.items()):
        ref = COMMIT + ":" + name
        size = subprocess.check_output(["git", "cat-file", "-s", ref], cwd=repo, timeout=10)
        require(len(size) <= 32 and int(size) == row["bytes"], "Immutable source size differs")
        child = subprocess.Popen(["git", "cat-file", "blob", ref], cwd=repo, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
        try:
            raw = child.stdout.read(row["bytes"] + 1)
            require(len(raw) == row["bytes"], "Immutable source exceeds declared size")
            require(child.wait(timeout=10) == 0 and sha(raw) == row["sha256"], "Immutable source digest differs")
        finally:
            child.stdout.close()
            if child.poll() is None:
                child.kill()
                child.wait()
        checked.append(row)
    return checked


class NoRedirect(HTTPRedirectHandler):
    def redirect_request(self, *_args, **_kwargs):
        raise ValueError("Publication redirect refused")


def fetch_one(name, expected, deadline):
    started = time.monotonic()
    row = {"path": name, "url": BASE + name, "started_at": datetime.now(timezone.utc).isoformat(), "status": "failed"}
    try:
        remaining = deadline - started
        require(remaining > 0, "Total request time bound exceeded")
        request = Request(row["url"], headers={"Accept-Encoding": "identity", "User-Agent": "OKF-DWP bounded publication verifier"})
        with build_opener(NoRedirect()).open(request, timeout=min(15, remaining)) as response:
            row.update(http_status=response.status, response_url=response.geturl())
            require(response.status == 200 and response.geturl() == row["url"], "Unexpected response status or URL")
            chunks, length = [], 0
            while length <= expected["bytes"]:
                require(time.monotonic() < deadline, "Total request time bound exceeded")
                chunk = response.read1(min(65536, expected["bytes"] + 1 - length))
                if not chunk:
                    break
                chunks.append(chunk)
                length += len(chunk)
            raw = b"".join(chunks)
        row.update(bytes=len(raw), sha256=sha(raw))
        require(len(raw) == expected["bytes"] and sha(raw) == expected["sha256"], "Published bytes differ from exact generated candidate")
        row["status"] = "matched"
        return row, raw
    except Exception as error:
        row["error_type"] = type(error).__name__
        # Avoid retaining arbitrary server response bodies or exception strings.
        return row, None
    finally:
        row["elapsed_ms"] = round((time.monotonic() - started) * 1000, 2)


class Links(HTMLParser):
    def __init__(self):
        super().__init__()
        self.ids = []
        self.links = []

    def handle_starttag(self, tag, attrs):
        attrs = dict(attrs)
        if "id" in attrs:
            self.ids.append(attrs["id"])
        if tag == "a" and "href" in attrs:
            self.links.append(attrs["href"])


def audit_links(materials):
    pages = {}
    for name, raw in materials.items():
        if name.endswith(".html"):
            page = Links()
            page.feed(raw.decode("utf-8"))
            pages[name] = page
    errors, per_page = [], []
    internal = external = 0
    for name, page in sorted(pages.items()):
        local = outside = 0
        for identifier in set(page.ids):
            if page.ids.count(identifier) > 1:
                errors.append({"page": name, "kind": "duplicate-id", "target": identifier})
        for href in page.links:
            parsed = urlsplit(urljoin(BASE + name, href))
            base = urlsplit(BASE)
            if parsed.scheme != base.scheme or parsed.netloc != base.netloc or not parsed.path.startswith(base.path):
                outside += 1
                continue
            local += 1
            target = unquote(parsed.path[len(base.path):])
            if not target or target.endswith("/"):
                target += "index.html"
            normalised = posixpath.normpath(target)
            if normalised != target or target not in materials:
                errors.append({"page": name, "kind": "missing-target", "target": href})
            elif parsed.fragment and target.endswith(".html") and unquote(parsed.fragment) not in pages[target].ids:
                errors.append({"page": name, "kind": "missing-fragment", "target": href})
        internal += local
        external += outside
        per_page.append({"path": name, "ids": len(page.ids), "internal_links": local, "external_links_not_fetched": outside})
    return {"html_pages": len(pages), "internal_links": internal, "external_links_not_fetched": external,
            "errors": errors, "pages": per_page,
            "scope": "Anchor hrefs and fragment IDs in received HTML; not external links, browser layout or assistive-technology acceptance."}


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--dwp-root", type=Path, required=True)
    parser.add_argument("--expected-manifest", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    directory(args.output.parent)
    # Refuse an existing file or symlink before network or source reads.
    fd = os.open(args.output, os.O_CREAT | os.O_EXCL | os.O_WRONLY | os.O_NOFOLLOW, 0o644)
    result = {"schema": "okf-dwp-learning-site-public-verification.v2", "source_commit": COMMIT,
              "base_url": BASE, "started_at": datetime.now(timezone.utc).isoformat(), "status": "failed",
              "requests": [], "limits": {"files": MAX_FILES, "manifest_bytes": MAX_MANIFEST,
              "member_bytes": MAX_MEMBER, "total_bytes": MAX_TOTAL, "parallel_requests": 4,
              "socket_operation_timeout_seconds": 15, "request_start_and_read_deadline_seconds": 180, "retries": 0},
              "limitations": ["Point-in-time public HTTP identity and internal-link audit; no browser, model or external-link verification.",
                              "No corpus downloads or private correspondence reads; only allowlisted public documentation source blobs."]}
    try:
        result["phase"] = "expected-manifest"
        expected_raw = bounded_read(args.expected_manifest, MAX_MANIFEST)
        expected = json.loads(expected_raw)
        sources, files = validate_manifest(expected)
        require(sum(row["bytes"] + 1 for row in files.values()) + len(expected_raw) + 1 <= MAX_TOTAL,
                "Response body reservations exceed aggregate byte bound")
        result.update(expected_manifest_sha256=sha(expected_raw), verifier_sha256=sha(bounded_read(__file__, MAX_MEMBER)),
                      source_pages=len(sources), listed_outputs=len(files))
        result["phase"] = "immutable-source-blobs"
        result["immutable_sources"] = verify_sources(args.dwp_root, sources)
        deadline = time.monotonic() + 180
        result["phase"] = "public-manifest"
        row, public_raw = fetch_one("site-manifest.json", {"bytes": len(expected_raw), "sha256": sha(expected_raw)}, deadline)
        result["requests"].append(row)
        require(public_raw is not None, "Public manifest identity did not match")
        public = json.loads(public_raw)
        _, actual_files = validate_manifest(public)
        materials = {"site-manifest.json": public_raw}
        result["phase"] = "public-files"
        with ThreadPoolExecutor(max_workers=4) as executor:
            pending = [(name, executor.submit(fetch_one, name, item, deadline)) for name, item in sorted(actual_files.items())]
            for name, future in pending:
                request_row, raw = future.result()
                result["requests"].append(request_row)
                if raw is not None:
                    materials[name] = raw
        require(all(row["status"] == "matched" for row in result["requests"]), "One or more public files failed verification")
        result["phase"] = "internal-links"
        result["link_audit"] = audit_links(materials)
        require(not result["link_audit"]["errors"], "Internal link audit failed")
        result["status"] = "passed"
        result["phase"] = "complete"
    except Exception as error:
        result["error_type"] = type(error).__name__
    finally:
        result["completed_at"] = datetime.now(timezone.utc).isoformat()
        result["matched_requests"] = sum(row["status"] == "matched" for row in result["requests"])
        result["transferred_bytes"] = sum(row.get("bytes", 0) for row in result["requests"])
        with os.fdopen(fd, "w") as stream:
            json.dump(result, stream, indent=2)
            stream.write("\n")
    print(json.dumps({key: result[key] for key in ("status", "source_commit", "matched_requests", "transferred_bytes")}))
    return 0 if result["status"] == "passed" else 1


if __name__ == "__main__":
    raise SystemExit(main())
